pkg/cloudagents/client.go

```diff
@@ -0,0 +1,159 @@
+	httpClient *http.Client
+	headers    map[string]string
+	logger     logger.Logger
+}
+
+// New returns a new Client with the given project URL, API key, and API secret.
+func New(opts ...ClientOption) (*Client, error) {
+	client := &Client{
+		logger: logger.GetLogger(),
+	}
+	for _, opt := range opts {
+		opt(client)
+	}
+	if client.projectURL == "" {
+		return nil, fmt.Errorf("project credentials are required")
+	}
+	agentClient, err := lksdk.NewAgentClient(client.projectURL, client.apiKey, client.apiSecret, twirp.WithClientHooks(&twirp.ClientHooks{
+		RequestPrepared: func(ctx context.Context, req *http.Request) (context.Context, error) {
+			client.setLivekitHeaders(req)
+			return ctx, nil
+		},
+	}))
+	if err != nil {
+		return nil, err
+	}
+	client.AgentClient = agentClient
+	client.agentsURL = client.getAgentsURL("")
+	if client.httpClient == nil {
+		client.httpClient = &http.Client{}
+	}
+	return client, nil
+}
+
+// CreateAgent creates a new agent by building from source.
+func (c *Client) CreateAgent(
+	ctx context.Context,
+	source fs.FS,
+	secrets []*lkproto.AgentSecret,
+	regions []string,
+	excludeFiles []string,
+) (*lkproto.CreateAgentResponse, error) {
+	resp, err := c.AgentClient.CreateAgent(ctx, &lkproto.CreateAgentRequest{
+		Secrets: secrets,
+		Regions: regions,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.uploadAndBuild(ctx,
+		resp.AgentId,
+		resp.PresignedUrl,
+		resp.PresignedPostRequest,
+		source,
+		excludeFiles,
+	); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// DeployAgent deploys new agent by building from source.
+func (c *Client) DeployAgent(
+	ctx context.Context,
+	agentID string,
+	source fs.FS,
+	secrets []*lkproto.AgentSecret,
+	excludeFiles []string,
+) error {
+	resp, err := c.AgentClient.DeployAgent(ctx, &lkproto.DeployAgentRequest{
+		AgentId: agentID,
+		Secrets: secrets,
+	})
+	if err != nil {
+		return err
+	}
+	if !resp.Success {
+		return fmt.Errorf("failed to deploy agent: %s", resp.Message)
+	}
+	return c.uploadAndBuild(ctx, agentID, resp.PresignedUrl, resp.PresignedPostRequest, source, excludeFiles)
+}
+
+// uploadAndBuild uploads the source and triggers remote build
+func (c *Client) uploadAndBuild(
+	ctx context.Context,
+	agentID string,
+	presignedUrl string,
+	presignedPostRequest *lkproto.PresignedPostRequest,
+	source fs.FS,
+	excludeFiles []string,
+) error {
+	if err := uploadSource(
+		source,
+		presignedUrl,
+		presignedPostRequest,
+		excludeFiles,
+	); err != nil {
+		return err
+	}
+	if err := c.build(ctx, agentID); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) getAgentsURL(serverRegion string) string {
+	agentsURL := c.projectURL
+	if strings.HasPrefix(agentsURL, "ws") {
+		agentsURL = strings.Replace(agentsURL, "ws", "http", 1)
+	}
+	if os.Getenv("LK_AGENTS_URL") != "" {
+		agentsURL = os.Getenv("LK_AGENTS_URL")
+	} else if !strings.Contains(agentsURL, "localhost") && !strings.Contains(agentsURL, "127.0.0.1") {
```

Comment:
Shall we overwrite `agentsURL` here? I think env var. has a lowest priority?

Reply:
good find. Will fix